src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,323 @@
+      acos(Mn::Math::dot(vec.normalized(), Mn::Vector3(0, 1, 0))))};
+  Mn::Vector3 crossProduct = Mn::Math::cross(vec, Mn::Vector3(0, 1, 0));
+  Mn::Quaternion vecRotation{Mn::Math::IdentityInit};
+
+  if (vec[0] == 0 && vec[2] == 0) {
+    crossProduct = Mn::Vector3(0, 1, 0);
+  }
+  vecRotation = vecRotation.rotation(-angle, crossProduct.normalized());
+  for (const auto& ind : conePositions) {
+    positions.push_back(
+        vecRotation.transformVector(ind * Mn::Vector3(0.02, 0.04, 0.02) +
+                                    Mn::Vector3(0, 0.025, 0)) +
+        mid);
+    colors.push_back(Mn::Color3(0, .3, 1));
+  }
+
+  for (const auto& norm : coneNormals) {
+    normals.push_back(norm);
+  }
+
+  for (const auto&& index : coneIndices) {
+    indices.push_back(sz + index[0]);
+  }
+
+  // render cylinder (arrow stem)
+  Mn::Trade::MeshData cylinderData = Mn::Primitives::cylinderSolid(1, 3, 1.0f);
+
+  sz = positions.size();
+  const auto&& cylinderPositions = cylinderData.positions3DAsArray();
+  const auto&& cylinderNormals = cylinderData.normalsAsArray();
+  const auto&& cylinderIndices = cylinderData.indices();
+
+  for (const auto& ind : cylinderPositions) {
+    positions.push_back(
+        vecRotation.transformVector(ind * Mn::Vector3(0.01, 0.03, 0.01) -
+                                    Mn::Vector3(0, 0.025, 0)) +
+        mid);
+    colors.push_back(Mn::Color3(0, .3, 1));
+  }
+
+  for (const auto& norm : cylinderNormals) {
+    Mn::Vector3 transformedNorm =
+        vecRotation.transformVector(norm * Mn::Vector3(0.01, 0.04, 0.01));
+    normals.push_back(transformedNorm);
+  }
+
+  for (const auto&& index : cylinderIndices) {
+    indices.push_back(sz + index[0]);
+  }
+}
+
+void VoxelGrid::generateMeshDataAndMeshGL(const std::string& gridName,
+                                          std::vector<Mn::UnsignedInt>& indices,
+                                          std::vector<Mn::Vector3>& positions,
+                                          std::vector<Mn::Vector3>& normals,
+                                          std::vector<Mn::Color3>& colors) {
+  // If the mesh already exists, replace it. Otherwise, create a new entry in
+  // the dict
+  Mn::Trade::MeshData positionMeshData{
+      Mn::MeshPrimitive::Triangles,
+      {},
+      indices,
+      Mn::Trade::MeshIndexData{indices},
+      {},
+      positions,
+      {Mn::Trade::MeshAttributeData{Mn::Trade::MeshAttribute::Position,
+                                    Cr::Containers::arrayView(positions)}}};
+  meshDataDict_[gridName] =
+      std::make_shared<Mn::Trade::MeshData>(Mn::MeshTools::interleave(
+          positionMeshData,
+          {Mn::Trade::MeshAttributeData{Mn::Trade::MeshAttribute::Color,
+                                        Cr::Containers::arrayView(colors)},
+           Mn::Trade::MeshAttributeData{Mn::Trade::MeshAttribute::Normal,
+                                        Cr::Containers::arrayView(normals)}}));
```

Comment:
Step 3, instead of all this, we can populate the mesh data directly:

```cpp
void VoxelGrid::generateMeshDataAndMeshGL(const std::string& gridName,
                                          Cr::Containers::Array<Mn::UnsignedInt>&& indexData, // note the &&s (!)
                                          Cr::Containers::Array<VoxelVertex>&& vertexData) {
  // need to make the index/attribute views first, before calling std::move() on the data
  Mn::Trade::MeshIndexData indices{indexData}; 
  Mn::Trade::MeshAttributeData positions{ 
      Mn::Trade::MeshAttribute::Position,
      Cr::Containers::stridedArrayView(vertexData).slice(&VoxelVertex::position)}; // view on just the position member
  Mn::Trade::MeshAttributeData colors{ 
      Mn::Trade::MeshAttribute::Color,
      Cr::Containers::stridedArrayView(vertexData).slice(&VoxelVertex::color)}; // view on just the color member
  Mn::Trade::MeshAttributeData normals{ 
      Mn::Trade::MeshAttribute::Normal,
      Cr::Containers::stridedArrayView(vertexData).slice(&VoxelVertex::normal)}; // view on just the normal member
      
  // now move the data into a MeshData, together with all metadata describing indices and attributes
  Mn::Trade::MeshData data{
      Mn::MeshPrimitive::Triangles,
      // the cast takes an Array<UnsignedInt> and turns it into an Array<char> without having to copy
      // everything again
      Containers::arrayAllocatorCast<char>(std::move(indexData)), indices,
      Containers::arrayAllocatorCast<char>(std::move(vertexData)), {positions, colors, normals}};
      
  // and save those, again move because the neither the Array nor MeshData are copyable
  meshDataDict_[gridName] =
      std::make_shared<Mn::Trade::MeshData>(std::move(data));
```

The comments are there mainly for you to understand better what's going on, feel free to drop them / rewrite as needed.

- [Docs for `arrayAllocatorCast()`](https://doc.magnum.graphics/corrade/namespaceCorrade_1_1Containers.html#a46c91612607a6ce5a823322d05666879)
- [Docs for the weird member `slice()`](https://doc.magnum.graphics/corrade/classCorrade_1_1Containers_1_1StridedArrayView.html#ababb10360346c92e016188824b54d61d)